src/gui/window/keyboard.rs

```diff
@@ -28,28 +37,169 @@ impl GuiKeyboard {
+                }
+                let color3 = Color32::from_rgb(20, 20, 20);
+                let top_left3 = Pos2::new(map_x(key.right), top);
+                let bottom_right3 = Pos2::new(map_x(key.right) - onepx, bottom);
+                let rect3 = Rect::from_min_max(top_left3, bottom_right3);
+                mesh.add_colored_rect(rect3, color3);
             }
         }
 
+        let bar_top_left = Pos2::new(rect.left(), rect.top());
+        let bar_bottom_right = Pos2::new(rect.right(), top);
+        let bar_rect = Rect::from_min_max(bar_top_left, bar_bottom_right);
+
+        mesh.add_colored_rect(bar_rect, *bar_color);
+
         for (i, key) in key_view.iter_visible_keys() {
             if key.black {
-                let top_left = Pos2::new(map_x(key.left), top);
-                let bottom_right = Pos2::new(map_x(key.right), black_bottom);
-
-                let rect = Rect::from_min_max(top_left, bottom_right);
                 let color = colors[i].map(map_color).unwrap_or(Color32::BLACK);
 
-                mesh.add_colored_rect(rect, color)
+                if color == Color32::BLACK {
+                    // Not pressed
+                    let black_key_bottom = md_height;
+
+                    let top_left1 = Pos2::new(map_x(key.left) - onepx, top);
+                    let bottom_right1 = Pos2::new(map_x(key.right) + onepx, black_bottom);
+                    let rect1 = Rect::from_min_max(top_left1, bottom_right1);
+                    let color1 = Color32::from_rgb(62, 62, 62);
+
+                    let top_left2 = Pos2::new(map_x(key.left) + onepx, top - black_key_overlap);
+                    let bottom_right2 =
+                        Pos2::new(map_x(key.right) - onepx, black_bottom - black_key_bottom);
+                    let rect2 = Rect::from_min_max(top_left2, bottom_right2);
+                    let color2 = Color32::from_rgb(24, 24, 24);
+
+                    mesh.add_colored_rect(rect1, color1);
+                    mesh.add_colored_rect(rect2, color2);
+                } else {
+                    // Pressed
+                    let top_left1 = Pos2::new(map_x(key.left), top);
+                    let bottom_right1 = Pos2::new(map_x(key.right), black_bottom);
+                    let rect1 = Rect::from_min_max(top_left1, bottom_right1);
+                    let color1 = Color32::from_rgb(color.r() / 2, color.g() / 2, color.b() / 2);
+
+                    let top_left2 = Pos2::new(map_x(key.left) + onepx, top);
+                    let bottom_right2 = Pos2::new(map_x(key.right) - onepx, black_bottom - onepx);
+                    let rect2 = Rect::from_min_max(top_left2, bottom_right2);
+
+                    mesh.add_colored_rect(rect1, color1);
+                    mesh.add_colored_rect(rect2, color);
+                }
             }
         }
 
         ui.painter().add(mesh);
     }
+
+    pub fn draw_empty(&mut self, ui: &mut Ui, key_view: &KeyboardView, bar_color: &Color32) {
```

Comment:
Same as with the window, this is quite a lot of duplicated code that ideally should be done together. You should be able to make an array of 256 None values and pass that into the draw function rather than making a new function entirely